jax/_src/prng.py

```diff
@@ -0,0 +1,475 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from functools import partial
+from typing import Callable, Iterator, NamedTuple
+import warnings
+
+import numpy as np
+
+from jax import lax
+from jax import core
+from jax import numpy as jnp
+from jax import tree_util
+from jax._src.api import jit, vmap
+from jax.config import config
+from jax.lib import xla_bridge
+from jax.lib import xla_client
+from jax.lib import cuda_prng
+from jax.interpreters import batching
+from jax.interpreters import xla
+from jax._src.pprint_util import pp, vcat
+from jax._src.util import prod
+
+
+UINT_DTYPES = {
+    8: jnp.uint8, 16: jnp.uint16, 32: jnp.uint32, 64: jnp.uint64}  # type: ignore[has-type]
+
+# -- PRNG implementation interface --
+
+class PRNGImpl(NamedTuple):
+  key_shape: core.Shape
+  seed: Callable
+  split: Callable
+  random_bits: Callable
+  fold_in: Callable
```

Comment:
can you more explicitly type the function signatures? Or document them here?

Reply:
Good call! Done.